osquery/sql/sqlite_util.cpp

```diff
@@ -459,41 +461,110 @@ Status QueryPlanner::applyTypes(TableColumns& columns) {
   return Status(0);
 }
 
-int queryDataCallback(void* argument, int argc, char* argv[], char* column[]) {
-  if (argument == nullptr) {
-    VLOG(1) << "Query execution failed: received a bad callback argument";
-    return SQLITE_MISUSE;
-  }
-
-  auto qData = static_cast<QueryData*>(argument);
-  Row r;
-  for (int i = 0; i < argc; i++) {
-    if (column[i] != nullptr) {
-      if (r.count(column[i])) {
-        // Found a column name collision in the result.
-        VLOG(1) << "Detected overloaded column name " << column[i]
-                << " in query result consider using aliases";
-      }
-      r[column[i]] = (argv[i] != nullptr) ? argv[i] : FLAGS_nullvalue;
+// Wrapper for legacy method until all uses can be replaced
+Status queryInternal(const std::string& query,
+                     QueryData& results,
+                     const SQLiteDBInstanceRef& instance) {
+  QueryDataJson jsonResults;
+Status status = queryInternal(query, jsonResults, instance);
+  for (rj::Document& obj : jsonResults) {
+    Row r;
+    for (rj::Value::ConstMemberIterator itr = obj.MemberBegin();
+         itr != obj.MemberEnd();
+         ++itr) {
+      r[itr->name.GetString()] = itr->value.GetString();
     }
+    results.push_back(r);
   }
-  (*qData).push_back(std::move(r));
-  return 0;
+  return status;
 }
 
-Status queryInternal(const std::string& q,
-                     QueryData& results,
+Status queryInternal(const std::string& query,
+                     QueryDataJson& results,
                      const SQLiteDBInstanceRef& instance) {
-  char* err = nullptr;
-  auto lock = instance->attachLock();
-  sqlite3_exec(instance->db(), q.c_str(), queryDataCallback, &results, &err);
+  sqlite3_stmt* pStmt{nullptr}; /* Statement to execute. */
+
+  int rc = SQLITE_OK; /* Return Code */
+  int rc2; /* Used to temporarily hold rc at end of big while loop */
+  const char* zLeftover; /* Tail of unprocessed SQL */
+  const char* zSql = query.c_str(); /* SQL to be processed */
+
+  /* The big while loop.  One iteration per statement */
+  while ((zSql[0] != 0) && (SQLITE_OK == rc)) {
+    auto lock = instance->attachLock();
+
+    rc = sqlite3_prepare_v2(instance->db(), zSql, -1, &pStmt, &zLeftover);
+    if (rc != SQLITE_OK || pStmt == nullptr) {
+      if (pStmt != nullptr) {
+        sqlite3_finalize(pStmt);
+      }
+      return Status(1, sqlite3_errmsg(instance->db()));
+    } else {
+      if (pStmt == nullptr) {
+        /* this happens for a comment or white-space */
+        zSql = zLeftover;
+        while (isspace(zSql[0])) {
+          zSql++;
+        }
+        continue;
+      }
+
+      rc = sqlite3_step(pStmt);
+      /* if we have a result set row... */
+      if (SQLITE_ROW == rc) {
+        // First collect the column names and types
+        int nCol = sqlite3_column_count(pStmt);
+        std::vector<std::string> colNames(nCol);
```

Comment:
I think it's a bug, reserve should be done instead of the resize

Reply:
This variable should also be snakecase :)